pkg/deviceclaimingserver/gateways/ttgc/ttgc.go

```diff
@@ -16,35 +16,198 @@
+			!bytes.Equal(profile.RootCa, loraPFProfile.RootCa) {
+			_, err := loraPFProfileClient.Update(ctx, &northboundv1.LoraPacketForwarderProfileServiceUpdateRequest{
+				Domain:                     u.client.Domain(ctx),
+				Group:                      profileGroup,
+				ProfileId:                  loraPFGetRes.ProfileId,
+				LoraPacketForwarderProfile: loraPFProfile,
+			})
+			if err != nil {
+				logger.WithError(err).Warn("Failed to update LoRa Packet Forwarder profile")
+				return err
+			}
+		}
+		loraPFProfileID = loraPFGetRes.ProfileId
+	}
+
+	// Upsert the Geolocation profile.
+	var (
+		geolocationProfileID []byte
+		geolocationProfile   = &northboundv1.GeolocationProfile{
+			ProfileName:     "on connect",
+			Shared:          true,
+			DisconnectedFor: durationpb.New(0),
+		}
+		geolocationProfileClient = northboundv1.NewGeolocationProfileServiceClient(u.client)
+	)
+	geolocationGetRes, err := geolocationProfileClient.GetByName(
+		ctx,
+		&northboundv1.GeolocationProfileServiceGetByNameRequest{
+			Domain:      u.client.Domain(ctx),
+			Group:       profileGroup,
+			ProfileName: geolocationProfile.ProfileName,
+		},
+	)
+	if err != nil {
+		if status.Code(err) != codes.NotFound {
+			logger.WithError(err).Warn("Failed to get geolocation profile")
+			return err
+		}
+		res, err := geolocationProfileClient.Create(ctx, &northboundv1.GeolocationProfileServiceCreateRequest{
+			Domain:             u.client.Domain(ctx),
+			Group:              profileGroup,
+			GeolocationProfile: geolocationProfile,
+		})
+		if err != nil {
+			logger.WithError(err).Warn("Failed to create geolocation profile")
+			return err
+		}
+		geolocationProfileID = res.ProfileId
+	} else {
+		geolocationProfileID = geolocationGetRes.ProfileId
+	}
+
+	// Update the gateway with the profiles.
+	_, err = gtwClient.Update(ctx, &northboundv1.GatewayServiceUpdateRequest{
+		GatewayId: eui.MarshalNumber(),
+		Domain:    u.client.Domain(ctx),
+		LoraPacketForwarderProfileId: &northboundv1.ProfileIDValue{
+			Value: loraPFProfileID,
+		},
+		GeolocationProfileId: &northboundv1.ProfileIDValue{
+			Value: geolocationProfileID,
+		},
+	})
+	if err != nil {
+		logger.WithError(err).Warn("Failed to update gateway with profiles")
+		return err
+	}
+
+	return nil
 }
 
 // Unclaim implements gateways.GatewayClaimer.
-func (TTGC) Unclaim(context.Context, types.EUI64, string) error {
-	return errUnimplemented.New()
+func (u *Upstream) Unclaim(ctx context.Context, eui types.EUI64) error {
+	gtwClient := northboundv1.NewGatewayServiceClient(u.client)
+	_, err := gtwClient.Unclaim(ctx, &northboundv1.GatewayServiceUnclaimRequest{
+		GatewayId: eui.MarshalNumber(),
+		Domain:    u.client.Domain(ctx),
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// IsManagedGateway implements gateways.GatewayClaimer.
+// This method always returns true.
+func (*Upstream) IsManagedGateway(context.Context, types.EUI64) (bool, error) {
+	return true, nil
 }
```

Comment:
This is fine for now but when (if) we add CUPS support to the GC, we need to make this dependent on the EUI range.

Reply:
That's why there's a `types.EUI64` argument and this method returns an error: we can differentiate EUI ranges (based on config) and even make an rpc call to figure this out.